### PR TITLE
DrivAerML: T_max=50 cosine schedule (longer cycles)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Champion uses T_max=30. Longer cosine periods (T_max=50) allow the optimizer to spend more time descending into each basin before the LR kicks back up. AirfRANS benefited significantly from T_max=50 vs T_max=10.

## Instructions

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 50 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group dm-tmax-sweep
```

## Reporting Requirements
- Best val `surface_rel_l2_pct` + epoch
- Test at best val checkpoint
- W&B run ID

## Baseline

**DrivAerML baselines:**
- val best: `3.997%` @ ep467 (PR #2898, W&B: `bht6h42t`, 4L/512d, AdamW lr=5e-4, T_max=30)
- test best: `6.244%` (PR #2648, 4L/320d, truncated eval — OLDER CONFIG)
- External targets: AB-UPT 3.82% | Transolver-3: 3.71%

Note: T_max=5 diverged (#2911). T_max=30 is champion. T_max=50 never tested at current 4L/512d config.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT himmel:

## Results

### DrivAerML: T_max=50 cosine schedule (4L/512d AdamW)

| Metric | Baseline (T_max=30) | T_max=50 | Delta |
|--------|---------------------|----------|-------|
| **Best val `surface_rel_l2_pct`** | 3.997% @ ep467 | **12.76% @ ep54** | +219% (worse) |
| Divergence epoch | Stable through 467 | **ep61** (grad explosion) | — |

**W&B run:** [oilrfmix](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/oilrfmix)

**Command:**
\`\`\`bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 50 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group dm-tmax-sweep \
  --agent himmel --wandb-name "himmel/dm-tmax-50"
\`\`\`

**Peak VRAM:** ~6.2 GB (GPU 0)

### What happened

**Hypothesis falsified.** T_max=50 does NOT benefit DrivAerML the way it benefited AirfRANS.

**Phase 1 — Healthy descent (ep1-58):** The model descended steadily through 5 cosine cycles, with each cycle's trough pushing deeper: 32.5% → 19.2% → ~16.4% → 15.7% → 12.8%. The descent envelope was clean and oscillation amplitude decreased cycle-over-cycle. Best val of **12.76% at epoch 54**.

**Phase 2 — Gradient explosion (ep61-62):** At epoch 61, grad norms exploded from 3.4 → 27.3 → 100.1. Loss jumped from 0.042 to 0.444. Val metric collapsed to 63.8%. No `--grad-clip` was set (matching the baseline config), but T_max=50 creates gradient instability that T_max=30 does not.

**Phase 3 — Partial recovery (ep63-66+):** The model partially recovered (63.8% → 32.9%) but this is essentially re-training from the epoch 4 level. The divergence erased ~50 epochs of progress.

**Key insight:** T_max=50 on DrivAerML creates longer periods of sustained high LR that eventually trigger catastrophic gradient explosions. The T_max=30 baseline survives 467+ epochs because the shorter cycles limit the duration of high-LR exposure. AirfRANS benefits from T_max=50 because its 2L/256d architecture is inherently more stable — the 4L/512d DrivAerML config has more parameters and deeper gradient paths, making it more susceptible to sustained high-LR instability.

Even before divergence, T_max=50's best (12.76%) at epoch 54 was 3.2x worse than the T_max=30 baseline (3.997%) at epoch 467. The slower cosine cycle means fewer descent opportunities per epoch, which is the opposite of what DrivAerML needs.

### Suggested follow-ups
- **T_max=50 + grad-clip 1.0:** The divergence might be preventable with gradient clipping. If clipping stabilizes the run, longer training could potentially match or beat T_max=30. This is the most informative next step.
- **T_max=40:** A middle ground between the champion T_max=30 and the unstable T_max=50. May capture some basin-depth benefit without divergence.
- **T_max=50 at lower LR (e.g. 3e-4):** Lower peak LR could prevent the gradient explosion while retaining the longer cycle benefit.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3080 because ADVISOR: Closing — T_max=50 is falsified for DrivAerML.

**Results:**
- Best val: 12.76% at ep54 — 3.19x worse than baseline 3.997%
- Gradient explosion at ep61 (val spiked to 63.80%)
- Run still recovering at ep71 (val=23.34%) — trajectory is uncompetitive

**Analysis:** Your explanation is correct — the 4L/512d architecture has deeper gradient paths that require shorter cosine cycles. T_max=50 provides only ~1.4 full cycles by ep70 vs ~2.3 for T_max=30, meaning the optimizer spends too long in high-LR regions. The AirfRANS→DrivAerML T_max transfer is falsified.

**Conclusion:** T_max=30 confirmed optimal for DM. historia (#3082) testing T_max=40 will complete the picture, but T_max>30 is clearly the wrong direction.

Good systematic work. Reassigning you to a new DrivAerML experiment..

**Comment by tcapelle** [comment]
STUDENT himmel:

## Updated Results (full run completed)

The training ran for **523 epochs** (360-min timeout). The model recovered from the epoch 61 divergence and found a deeper basin:

| Metric | Baseline (T_max=30) | T_max=50 (final) | Delta |
|--------|---------------------|-------------------|-------|
| **Best val `surface_rel_l2_pct`** | 3.997% @ ep467 | **11.33% @ ep124** | +183% (worse) |
| **Best test `surface_rel_l2_pct`** | 6.244% (older cfg) | **11.90% @ ep124** | — |
| Total epochs | 467 | 523 | — |
| Divergence events | None | ep61 (grad_norm→100), ep523 (grad_norm→267k) | — |

**Key update:** The model recovered from the ep61 divergence and continued improving through subsequent cycles, reaching 11.33% at epoch 124. However, a second, much larger gradient explosion occurred by the final epoch (grad_norm = 266,724 at ep523), confirming T_max=50's instability on the 4L/512d DrivAerML config.

**Conclusion unchanged:** T_max=50 is strictly worse than T_max=30 on DrivAerML (11.33% vs 3.997%), and suffers from repeated gradient explosions without grad-clip. The hypothesis is falsified.
